# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: This would fall into the catogory of a classification problem. We'd like to take a number of student attributes as inputs and output need/not need intervention**

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [2]:
# Calculate number of students
n_students = student_data.shape[0]

# Calculate number of features
n_features = student_data.shape[1]-1

#Calculate passing and failing students
n_passed, n_failed = student_data.passed.value_counts()

# Calculate graduation rate
grad_rate = float(n_passed)/(float(n_passed+n_failed))

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 0.67%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [3]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [4]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [5]:
from sklearn.model_selection import train_test_split

# Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size=95, train_size=300, random_state=616, shuffle=True)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: I would investigate Decision Trees, Random Forest, and K-nearest neigbors algorithms. Below are the reasons.**

Decision Trees:

Decision trees algorithm acts on the data set and choose the attributes that splits the data the most and then split the data accordingly. The model can produce complicated dicision boundaries that handles non-linearly seperatable data. Note that each piece of the decision boundary is parallel to the axes. The model trains very fast and can be scaled to large data set.

The downside of the decision trees algorithm includes overfitting and non-stable when data contains large amount feature space and not many features can cut the data clearly. To mitigate the overfitting issue, we can tune parameter set, prune the decision tree, and/or use random forest ensemble method.

Decision trees will likly be a good model to be applied to this problem because I anticipate there will be a number of attributes of the student that give good seperation on the data set, including address, famsize, reason, travel_time, famrel etc.

Random Forest:

Randome forest algorithm is an ensemble based algorithm extented from the decision tree algorithm. it takes a subset of the dataset and feature space and perform a decision tree algorithm on it. Generally 2/3 of the data set is used for training and the other 1/3 is used for cross-validation. The tree uses m features (generally m=sqrt of total number of features). The process is then repeated K times to produce K decision trees. In the end all decision trees are combined together by **weighted voting?**. https://www.quora.com/What-is-a-random-forest

Random forest algorithm is likly a good candiate for our problem on hand because we have 48 features, a decision tree without parameter tuning is likly to overfit. Therefore, it's worth looking into whether only selecting a subset of feature set for model training would mitigate the issue of overfitting

K-nearest neighbor

K-nearest neighbor is a type of classification where the data points are clustered based on the "distance" function defined. Usually the distance is taken as the euclidean distance (i.e. z = sqrt(x^2+y^2)), but it can be in any format. A new point is clssified based on how "close" it is to the existing clusters.
https://www.quora.com/How-would-you-explain-the-Nearest-Neighborhood-Classification-intuitively

Characteristics of the KNeighbor include 1) it's a fast train but slow prediction algorithm, unlike decision trees which is slow train but fast prediction. 2) This algorithm is capable of making multi-class classification out of the box.



### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [6]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [7]:
# Import the three supervised learning models from sklearn
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

# TODO: Initialize the three models
clf_A = DecisionTreeClassifier()
# parameters: max_depth, min_samples_split=2, min_samples_leaf=1
clf_B = RandomForestClassifier()
# parameters: n_estimators=10 (number of trees), max_depth, min_samples_split=2
# min_samples_leaf=1, max_features = "auto" (sqrt(n_features))
clf_C = KNeighborsClassifier()
# parameters: n_neighbors=5

# Set up the training set sizes
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

# Execute the 'train_predict' function for each classifier and each training set size
for model in ['clf_A','clf_B','clf_C']:
    model = locals()[model]
    for size in ['100','200','300']:
        X_data = locals()['X_train_'+size]
        y_data = locals()['y_train_'+size]
        train_predict(model, X_data, y_data, X_test, y_test)

    


Training a DecisionTreeClassifier using a training set size of 100. . .
Trained model in 0.0044 seconds
Made predictions in 0.0005 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0005 seconds.
F1 score for test set: 0.7941.
Training a DecisionTreeClassifier using a training set size of 200. . .
Trained model in 0.0031 seconds
Made predictions in 0.0011 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0007 seconds.
F1 score for test set: 0.6341.
Training a DecisionTreeClassifier using a training set size of 300. . .
Trained model in 0.0037 seconds
Made predictions in 0.0005 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0009 seconds.
F1 score for test set: 0.7538.
Training a RandomForestClassifier using a training set size of 100. . .
Trained model in 0.0308 seconds
Made predictions in 0.0018 seconds.
F1 score for training set: 0.9926.
Made predictions in 0.0016 seconds.
F1 score for test set: 0.7606.
Training a RandomForestClassifie

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Decision Trees**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |        0.0026        |       0.0007        |      1.0000      |      0.7704    |
| 200               |        0.0134        |       0.0008        |      1.0000      |      0.6281    |
| 300               |        0.0035        |       0.0010        |      1.0000      |      0.7576    |

** Random Forest**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100 | 0.0287|0.0014|0.9851|0.7910|
| 200 | 0.0319|0.0014|0.9889|0.8029|
| 300 | 0.0281|0.0014|0.9926|0.7353|

** KNeighbors**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100    |0.0008|0.0016|0.8079|0.7973|
| 200    |0.0010|0.0023|0.8097|0.7887|
| 300    |0.0013|0.0031|0.8533|0.7852|

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: The best model is decision tree model, based on the potential for F1 score improvment and run time.**

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: The decision tree classifier used here looks for features that best seperates the data into two groups. The features are the attributes of each person (i.e. sex, age, education years etc.), the two groups are income level greater than 50K and less than 50K. The algorithm works similar to the game 20 questions. Suppose we only have two features, age and education years. The best strategy would be ask whether a person is over 30 years old or not, if everyone who earn over 50K are over 30 years old. The next question is to ask whether a person received over 10 years of education, where supposly 10 years is a good seperator for people's income level. As a result, the model predicts everyone who is older than 30 and have received more than 10 years of education to have income more than 50K.**

**This model works the best if there are a number of features in the data set that gives good seperation of the data, which seems to be the case in our problem on hand. Therefore, we want to further fine tuning the model on our data set.
**

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.GridSearchCV`](http://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [8]:
from sklearn.metrics import make_scorer
from sklearn.model_selection import GridSearchCV, ShuffleSplit

# Create the parameters list you wish to tune
parameters1 = {'n_neighbors':[3,5,7,9,11,13]}
parameters2 = {'n_estimators':[8,11,14,17,20],'max_depth':[3,5,7,9,11,13],\
                             'min_samples_leaf':[5,7,10,13],'min_samples_split':[2,6,10,14,18]}
parameters3 = {'max_depth':[3,5,7,9,11,13],'min_samples_leaf':[5,7,10,13],'min_samples_split':[2,6,10,14,18]}

# Initialize the classifier
clf1 = KNeighborsClassifier()
clf2 = RandomForestClassifier()
clf3 = DecisionTreeClassifier()

# Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score,pos_label='yes')

# Create cross-validation sets from the training data
cv_sets = ShuffleSplit(n_splits = 5, test_size = 0.20, random_state = 16)

# Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj1 = GridSearchCV(estimator=clf1,param_grid=parameters1,scoring=f1_scorer,cv=cv_sets)
grid_obj2 = GridSearchCV(estimator=clf2,param_grid=parameters2,scoring=f1_scorer,cv=cv_sets)
grid_obj3 = GridSearchCV(estimator=clf3,param_grid=parameters3,scoring=f1_scorer,cv=cv_sets)

# Fit the grid search object to the training data and find the optimal parameters
grid_fit1 = grid_obj1.fit(X_train,y_train)
grid_fit2 = grid_obj2.fit(X_train,y_train)
grid_fit3 = grid_obj3.fit(X_train,y_train)

# Get the estimator
clf_opt_1 = grid_fit1.best_estimator_
clf_opt_2 = grid_fit2.best_estimator_
clf_opt_3 = grid_fit3.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned KNeighbor has a training F1 score of {:.4f}.".format(predict_labels(clf_opt_1, X_train, y_train))
print "Tuned KNeighbor has a testing F1 score of {:.4f}.".format(predict_labels(clf_opt_1, X_test, y_test))
print "Tuned KNeighbor has the following parameters",grid_fit1.best_params_
print "Tuned Random Forest has a training F1 score of {:.4f}.".format(predict_labels(clf_opt_2, X_train, y_train))
print "Tuned Random Forest has a testing F1 score of {:.4f}.".format(predict_labels(clf_opt_2, X_test, y_test))
print "Tuned Random Forest has the following parameters",grid_fit2.best_params_
print "Tuned Decision Tree has a training F1 score of {:.4f}.".format(predict_labels(clf_opt_3, X_train, y_train))
print "Tuned Decision Tree has a testing F1 score of {:.4f}.".format(predict_labels(clf_opt_3, X_test, y_test))
print "Tuned Decision Tree has the following parameters",grid_fit3.best_params_

Made predictions in 0.0081 seconds.
Tuned KNeighbor has a training F1 score of 0.8164.
Made predictions in 0.0030 seconds.
Tuned KNeighbor has a testing F1 score of 0.8243.
Tuned KNeighbor has the following parameters {'n_neighbors': 13}
Made predictions in 0.0024 seconds.
Tuned Random Forest has a training F1 score of 0.8511.
Made predictions in 0.0022 seconds.
Tuned Random Forest has a testing F1 score of 0.8158.
Tuned Random Forest has the following parameters {'min_samples_split': 18, 'n_estimators': 17, 'max_depth': 11, 'min_samples_leaf': 7}
Made predictions in 0.0007 seconds.
Tuned Decision Tree has a training F1 score of 0.8296.
Made predictions in 0.0007 seconds.
Tuned Decision Tree has a testing F1 score of 0.8194.
Tuned Decision Tree has the following parameters {'min_samples_split': 6, 'max_depth': 3, 'min_samples_leaf': 7}


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: The best algorithm for this problem, judged from both F1 score and running time is Decision tree. The F1 scores for training and testing are 0.829 and 0.819 respectively, indicating the model is not overfitting. The train and prediction time is one order of magnitude faster compare to the other two algorithms. Compare to untuned decision tree model, we successfully midigated the overfitting issue and achieved a 0.06 improvement in the F1 score.**

Recover the list of most important features.

In [23]:
feature_imp = {}
feature_list = X_train.columns.values

for i, item in enumerate(clf_opt_3.feature_importances_):
    if item > 0:
        feature_imp[feature_list[i]] = item
        
print feature_imp

{'failures': 0.6919898340152789, 'sex_F': 0.08485267701926302, 'Fjob_services': 0.0777007892530933, 'age': 0.10184911390705724, 'famrel': 0.043607585805307464}


Note five features have highest impact on the prediction:

1) number of past class failures

2) student's age

3) whether the student is a female

4) whether father's job is in civil service (admin or police)

5) quality of family relationships

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.